### PR TITLE
Misc. fixes for tests (locally+AppVeyor)

### PR DESCRIPTION
**Pull request by e2**
- removed test that didn't test anything (and failed on my setup)
- unset the BSD_ECHO param for zsh calls (disabled on my system)
- fix missing 'cmd' variable (AppVeyor failures)
- fix extraneous param to delete (AppVeyor failures)
- avoid checking for 'cli', because it's part of mono-runtime (Linux)
- fix block failures handling (nil on Windows/AppVeyor)

### CODE REVIEW

**Comment by maxmeyer** [comment]
@e2 Thanks a lot for your contribution. See my comments attached.

File: features/api/command/run.feature [hunk 1]
```diff
@@ -60,13 +60,13 @@ Feature: Run command
     Then the specs should all pass
 
   Scenario: Non-existing executable
-    Given a file named "bin/cli" does not exist
+    Given a file named "bin/cli-app" does not exist
```
Comment (maxmeyer) [inline features/api/command/run.feature]: `bin/cli` != `bin/cli-app` -> should be just `bin/cli`.

File: features/api/command/run.feature [hunk 2]
```diff
@@ -60,13 +60,13 @@ Feature: Run command
...
     And a file named "spec/run_spec.rb" with:
     """ruby
     require 'spec_helper'
 
     RSpec.describe 'Find path for command', :type => :aruba do
-      it { expect { run('cli') }.to raise_error Aruba::LaunchError, /Command "cli" not found in PATH-variable/ }
+      it { expect { run('cli-app') }.to raise_error Aruba::LaunchError, /Command "cli-app" not found in PATH-variable/ }
```
Comment (maxmeyer) [inline features/api/command/run.feature]: same here.

File: features/configuration/home_directory.feature [hunk 3]
```diff
@@ -43,24 +43,6 @@ Feature: Configure the home directory to be used with aruba
     The default value is "."
     """
 
-  Scenario: Set to aruba's working directory
-    Given a file named "features/support/aruba.rb" with:
```
Comment (maxmeyer) [inline features/configuration/home_directory.feature]: Why did you remove that test?

File: features/steps/command/shell.feature [hunk 4]
```diff
@@ -72,7 +72,7 @@ Feature: Running shell commands
       Scenario: Running zsh commands
         When I run the following commands with `zsh`:
         \"\"\"bash
-        echo "Hello \c"
+        echo -e "Hello \c"
```
Comment (maxmeyer) [inline features/steps/command/shell.feature]: `echo` vs `echo -e` are two different things.

File: spec/aruba/aruba_path_spec.rb [hunk 5]
```diff
@@ -98,6 +98,6 @@
     before(:each) { FileUtils.mkdir_p File.dirname(new_path) }
     before(:each) { File.open(new_path, 'w') { |f| f.print 'a' } }
 
-    it { expect(path.blocks).to be > 0 }
+    it { expect(path.blocks || 123).to be > 0 }
```
Comment (maxmeyer) [inline spec/aruba/aruba_path_spec.rb]: Why did you do that? That looks more like a work around.

File: lib/aruba/processes/spawn_process.rb [hunk 6]
```diff
@@ -83,7 +83,7 @@ def start
             sleep startup_wait_time
           end
         rescue ChildProcess::LaunchError => e
-          raise LaunchError, "It tried to start #{cmd}. " + e.message
+          raise LaunchError, "It tried to start #{commandline}. " + e.message
```
Comment (maxmeyer) [inline lib/aruba/processes/spawn_process.rb]: What's the intention behind this change?

File: features/api/command/run.feature (hunk 1 above)
Comment (e2) [inline features/api/command/run.feature]: I'm already confused. I'd call it "bin/nonexisting" or something more suggestive like that. Is that ok?

File: features/configuration/home_directory.feature (hunk 3 above)
Comment (e2) [inline features/configuration/home_directory.feature]: Because of the final test: testing if the path begins with "/home", which makes it no different from the other two tests (where the paths are fixed strings). On my setup, /home is symlinked, so this doesn't work anyway. I'm not sure how this method is testing the home directory method.

File: features/steps/command/shell.feature (hunk 4 above)
Comment (e2) [inline features/steps/command/shell.feature]: I have the following in my `~/.zshenv`: `setopt BSD_ECHO`

File: spec/aruba/aruba_path_spec.rb (hunk 5 above)
Comment (e2) [inline spec/aruba/aruba_path_spec.rb]: As I described in the commit: it can be nil if it isn't supported (it isn't on AppVeyor apparently - perhaps a network filesystem or something).

File: lib/aruba/processes/spawn_process.rb (hunk 6 above)
Comment (e2) [inline lib/aruba/processes/spawn_process.rb]: The variable `cmd` doesn't exist. The exception happens on AppVeyor, so the error is triggered there.

Comment (e2) [inline lib/aruba/processes/spawn_process.rb]: (`cmd` is passed to the init method - and not to this one. Probably the init was split into two methods some time ago.)

Comment (e2) [inline lib/aruba/processes/spawn_process.rb]: Oh, and after this fix, the affected tests on AppVeyor actually show the real errors now (instead of triggering the missing variable error).

**Comment by e2** [comment]
> Use the PR Template next time

This is about fixing failing tests, so exceptions are allowed. If tests aren't all green, something is very wrong already, and patching master is critical. (Otherwise all other PRs will be failing anyway - even if just on the desktops and not on Travis).

> Please, make separate commits for each change. Given the current way it's hard to see why you changed the code.

I'm fixing broken tests I didn't break, so I'm assuming I'm the least knowledgeable about them.

I can explain the changes, so just ask. (I though the commit message was enough).

I'm making the changes to fix the test, so the real question is: why were they failing and since when? What changed that caused them to fail? Why are broken AppVeyor tests ignored?

I'm patching with the best of my knowledge - knowing probably nothing more than you are by seeing these patches.

> I guess it's the test for the home variable? Correct? Can you add the output instead?

No, it's useless compared to the other two tests. Technically, it should just test if the variable is set or not. If the test is supposed to test contracts, one tests isn't enough, while additional steps would have to be added anyway. The test's description doesn't suggest anything meaningful either. Besides, the fact that it doesn't work on my symlinked home folder means it should be replaced with a new test with a clear intent. Again, one test isn't enough to be thorough, and if I can't justify adding multiple tests, it doesn't make a difference to just "avoid removing one test".

File: features/api/command/run.feature (hunk 1 above)
Comment (dg-ratiodata) [inline features/api/command/run.feature]: Na... The problem is this:

1. The fixture contains a default executable called `bin/cli`
2. We need to rename that file and each test using it, say to `bin/aruba-test-cmd` - this should avoid further conflicts with existing executables
3. I would prefer to stick to `Given a file named "bin/aruba-test-cmd" does not exist`, so we can still use a default command in our fixture and properly set state. We also explicitly say, that we asume, that such a file does not exist.

File: features/steps/command/shell.feature (hunk 4 above)
Comment (dg-ratiodata) [inline features/steps/command/shell.feature]: I think it's more a problem of test isolation. We/I wrongly assumed some shell - bash, zsh - configuration. But this may interfere with ones shell's setup - zshrc, bashrc.

I think it makes more sense, to explilcitly use a configuration for each shell to make sure, they're configured correctly.

Options I see:
1. Add a wrapper for `sh`, `bash`, `zsh` which uses it's only configuration file instead using the one of the user
2. Set the path to the configuration file via environment variable

What do you think?

File: spec/aruba/aruba_path_spec.rb (hunk 5 above)
Comment (dg-ratiodata) [inline spec/aruba/aruba_path_spec.rb]: As far as I know `File.blocks` and the like are not supported on windows. We never made it to the point where we made the specs aware of different operating systems/filesystems - we also started a discussion about that on [gitter](https://gitter.im/cucumber/aruba?at=56f8464de4a8384a1bbc6b3e) some days ago. What I think we need to do is, to fix the specs properly by only testing things on platforms where they are supposed to work.

File: lib/aruba/processes/spawn_process.rb (hunk 6 above)
Comment (dg-ratiodata) [inline lib/aruba/processes/spawn_process.rb]: Yeah, I think I changed that to `commandline`. You're right.

**Comment by dg-ratiodata** [comment]
Hey, thanks for the fast replies.

> I'm fixing broken tests I didn't break, so I'm assuming I'm the least knowledgeable about them.

I really like that list:

> * removed test that didn't test anything (and failed on my setup)
> * unset the BSD_ECHO param for zsh calls (disabled on my system)
> * fix missing 'cmd' variable (AppVeyor failures)
> * fix extraneous param to delete (AppVeyor failures)
> * avoid checking for 'cli', because it's part of mono-runtime (Linux)
> * fix block failures handling (nil on Windows/AppVeyor)

And I thought that all these single things would have been each a good description for a commit. So I could have seen, what change fixed what.

> I'm making the changes to fix the test, so the real question is: why were they failing and since when? What changed that caused them to fail? Why are broken AppVeyor tests ignored?

Simple: I added `AppVeyor.yml` some time ago, when we started to fix windows problems. But then the involved developers "vanished" and I'm not a windows developer. If users are interested in windows support they need to contribute. That's open source from my understanding.

> No, it's useless compared to the other two tests. 

What makes it worse, it might even fail on windows. So removing it is ok I think.

File: features/api/command/run.feature (hunk 1 above)
Comment (e2) [inline features/api/command/run.feature]: I'll need a separate PR for this. Aruba shouldn't even be using zsh in tests, because:

1. It hides the intention of what is being tested here (I'm still unsure, but I'm just guessing it's the path to an app, really)
2. Too much of a barrier to maintain (how many people will know the BSD_ECHO option?)
3. Doesn't work on Windows and will need to be revised (PITA)

(I commented on Gitter [here](https://gitter.im/cucumber/aruba?at=56fbed02e4a8384a1bbd47ae)

**Comment by e2** [comment]
I'm closing this, because everything mentioned is no longer relevant due to other PRs.
